### PR TITLE
[0.63] Allow swapping to System Chakra at runtime.

### PR DESCRIPTION
**Pull request by JunielKatarn**
CC @weidezhong1.

Enables using loading Chakra.dll instead of ChakraCore.dll for specialized scenarios without adding new DLLs.
(Note, this branch is based on `0.63-stable` in order to validate the non-JSI Chakra executor and the current RNW version used by MS Office).

- Refactors `ChakraRuntime` as an abstract class, allowing Chakra variant affinity to be resolved at runtime.
- Introduces runtime option `ForceSystemChakra`.

Current issues:
- ~~Fails `RNTesterIntegrationTests::Logging`.~~
  - **Solved**. This was a missing test fix backport.
- ~~If `ForceSystemChakra` is not set, integration tests crash due to a "binary mismatch".~~
  - **Solved**. Chakra.dll also needs to be explicitly delay-loaded.

**NOTE** This change will be applied to `master` first, then backported to `0.64-stable` and `0.63-stable`.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7479)

### CODE REVIEW

File: vnext/Desktop.IntegrationTests/RNTesterIntegrationTests.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@ std::wstring ToString<TestStatus>(const TestStatus &status) {
 
 TEST_MODULE_INITIALIZE(InitModule) {
   Microsoft::React::SetRuntimeOptionBool("WebSocket.AcceptSelfSigned", true);
+  Microsoft::React::SetRuntimeOptionBool("ForceSystemChakra", true);
```
Comment (JunielKatarn) [inline vnext/Desktop.IntegrationTests/RNTesterIntegrationTests.cpp:22]: TODO: Remove before submitting.

File: vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp [hunk 2]
```diff
@@ -86,6 +96,10 @@ bool ChakraRuntime::evaluateSerializedScript(
...
 }
 
+std::unique_ptr<facebook::jsi::Runtime> MakeSystemChakraRuntime(ChakraRuntimeArgs &&args) noexcept {
+  return std::make_unique<SystemChakraRuntime>(std::move(args));
+}
+
 } // namespace Microsoft::JSI
 
-#endif
+//#endif
```
Comment (JunielKatarn) [inline vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp:105]: Remove!

File: vnext/Shared/OInstance.cpp [hunk 3]
```diff
@@ -370,6 +371,10 @@ InstanceImpl::InstanceImpl(
       Microsoft::ReactNative::PackagerConnection::CreateOrReusePackagerConnection(*m_devSettings);
     }
 
+    if (Microsoft::React::GetRuntimeOptionBool("ForceSystemChakra")) {
```
Comment (JunielKatarn) [inline vnext/Shared/OInstance.cpp:374]: For reviewers: Is this an adequate location to set the JSI override conditionally?

File: vnext/Chakra/ChakraHelpers.cpp [hunk 4]
```diff
@@ -62,13 +63,14 @@ class ChakraVersionInfo {
...
 #if !defined(CHAKRACORE_UWP)
-    // This code is win32 only at the moment. We will need to change this
-    // line if we want to support UWP.
-    constexpr wchar_t chakraDllName[] = L"ChakraCore.dll";
-
     auto freeLibraryWrapper = [](void *p) { FreeLibrary((HMODULE)p); };
     HMODULE moduleHandle;
-    if (!GetModuleHandleExW(0, chakraDllName, &moduleHandle)) {
+    // This code is win32 only at the moment. We will need to change these
+    // linel if we want to support UWP.
```
Comment (vmoroz) [inline vnext/Chakra/ChakraHelpers.cpp:69]: >linel [](start = 7, length = 5)

typo

Comment (JunielKatarn) [reply vnext/Chakra/ChakraHelpers.cpp:69]: Good catch. It was a copy/paste block.

File: vnext/Desktop.DLL/pch.cpp [hunk 5]
```diff
@@ -2,3 +2,29 @@
 // Licensed under the MIT License.
 
 #include "pch.h"
+
+#include <RuntimeOptions.h>
```
Comment (vmoroz) [inline vnext/Desktop.DLL/pch.cpp:6]: This is the wrong file to put any code in. Th purpose of thius file must be only to generate the pch file. Please create a new .cpp file. E.g. DllMain or DelayLoadHook or something else along these lines.

File: vnext/Desktop.DLL/pch.cpp [hunk 6]
```diff
@@ -2,3 +2,29 @@
...
+
+    case dliStartProcessing:
+    case dliNotePreGetProcAddress:
+    case dliFailLoadLib:
+    case dliFailGetProc:
+    case dliNoteEndProcessing:
+      break;
+  }
+
+  return NULL;
```
Comment (vmoroz) [inline vnext/Desktop.DLL/pch.cpp:27]: >NULL [](start = 9, length = 4)

nullptr?

File: vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp [hunk 7]
```diff
@@ -1,40 +1,50 @@
...
+#if defined(USE_EDGEMODE_JSRT) && !defined(CHAKRACORE)
+/*static*/ void ChakraRuntime::initRuntimeVersion() noexcept {}
+#endif
+
+SystemChakraRuntime::SystemChakraRuntime(ChakraRuntimeArgs &&args) noexcept : ChakraRuntime(std::move(args)) {
+  this->Init();
+}
+
+void SystemChakraRuntime::setupNativePromiseContinuation() noexcept {
   // NOP
```
Comment (vmoroz) [inline vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp:25]: >// NOP [](start = 2, length = 6)

Please note that it is not NOP in 0.64 where we did a number of changes to these classes to make sure that all JSI tests are passing. We need to be very careful moving these changes to newer RNW versions.

File: vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp [hunk 8]
```diff
@@ -1,40 +1,50 @@
...
-}
-
-std::unique_ptr<const facebook::jsi::Buffer> ChakraRuntime::generatePreparedScript(
+std::unique_ptr<const facebook::jsi::Buffer> SystemChakraRuntime::generatePreparedScript(
     const std::string &sourceURL,
     const facebook::jsi::Buffer &sourceBuffer) noexcept {
   const std::wstring scriptUTF16 =
       Microsoft::Common::Unicode::Utf8ToUtf16(reinterpret_cast<const char *>(sourceBuffer.data()), sourceBuffer.size());
 
+#ifdef CHAKRACORE
```
Comment (vmoroz) [inline vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp:43]: It looks strange here - the file supposed to be about the System Chakra.

Comment (JunielKatarn) [reply vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp:43]: RNW Desktop always compiles with `CHAKRACORE` defined, yet, it needs to be able to instantiate this class if `ForceSystemChakra` is defined.

Hence, this code needs to be compiled with `CHAKRACORE` on.

File: vnext/JSI/Shared/ChakraObjectRef.cpp [hunk 9]
```diff
@@ -128,13 +129,17 @@ ChakraObjectRef GetPropertyId(const std::string_view &utf8) {
...
-  // We use a #ifdef here because we can avoid a UTF-8 to UTF-16 conversion
-  // using ChakraCore's JsCreatePropertyId API.
+// We use a #ifdef here because we can avoid a UTF-8 to UTF-16 conversion
+// using ChakraCore's JsCreatePropertyId API.
 #ifdef CHAKRACORE
-  JsPropertyIdRef id = JS_INVALID_REFERENCE;
-  VerifyChakraErrorElseThrow(JsCreatePropertyId(utf8.data(), utf8.length(), &id));
-  return ChakraObjectRef(id);
-
+  if (!Microsoft::React::GetRuntimeOptionBool("ForceSystemChakra")) {
```
Comment (vmoroz) [inline vnext/JSI/Shared/ChakraObjectRef.cpp:135]: >if (!Microsoft::React::GetRuntimeOptionBool("ForceSystemChakra")) { [](start = 2, length = 67)

Is it to enable the FeatureGate? I.e. it is the reason that we completely did not remove the whole #ifdef, right?

File: vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp (hunk 7 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/ChakraJsiRuntime_edgemode.cpp:25]: Definitely. I will manually port this changes to master once this review has been approved, and do the normal backport cycle.

I started the review here to be able to validate Office in parallel.

File: vnext/JSI/Shared/ChakraObjectRef.cpp (hunk 9 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/ChakraObjectRef.cpp:135]: Correct. The conditional compilation is still necessary in many spots.

The general idea is:
- Be able to compile RNW Desktop against ChakraCore includes and import libs.
- Be able to swap from ChakraCore to System Chakra at runtime.
- Leave MSRN (UWP) intact.

File: vnext/JSI/Shared/ChakraRuntime.h [hunk 10]
```diff
@@ -321,22 +307,23 @@ class ChakraRuntime : public facebook::jsi::Runtime {
...
   static std::once_flag s_runtimeVersionInitFlag;
   static uint64_t s_runtimeVersion;
 
   // Arguments shared by the specializations
   ChakraRuntimeArgs m_args;
 
-  JsRuntimeHandle m_runtime;
   ChakraObjectRef m_context;
 
   // Note: For simplicity, We are pinning the script and serialized script
```
Comment (vmoroz) [inline vnext/JSI/Shared/ChakraRuntime.h:329]: >// Note: [](start = 2, length = 8)

Should we move this comment too?

File: vnext/Shared/ChakraCoreRuntime.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+  bool evaluateSerializedScript(
+      const facebook::jsi::Buffer &scriptBuffer,
+      const facebook::jsi::Buffer &serializedScriptBuffer,
+      const std::string &sourceURL) override;
+
+#pragma endregion
+
+ private:
+  // core-only??
```
Comment (vmoroz) [inline vnext/Shared/ChakraCoreRuntime.h:42]: >// core-only?? [](start = 2, length = 14)

ChakraCore and system Chakra use different debuggers. So, the debugger initializaiton could be a little bit different between the two. It wonder if it is soething that can be abstracted between them at some level.

File: vnext/JSI/Shared/ChakraRuntime.h [hunk 12]
```diff
@@ -153,9 +140,29 @@ class ChakraRuntime : public facebook::jsi::Runtime {
     return m_args;
   }
 
- private:
   void VerifyJsErrorElseThrow(JsErrorCode error);
 
+  // These three functions only performs shallow copies.
+  facebook::jsi::Value ToJsiValue(ChakraObjectRef &&ref);
```
Comment (asklar) [inline vnext/JSI/Shared/ChakraRuntime.h:146]: typo

File: vnext/Shared/ChakraCoreRuntime.h (hunk 11 above)
Comment (JunielKatarn) [reply vnext/Shared/ChakraCoreRuntime.h:42]: From the original implementation, it looks like the debugging classes for System Chakra were defined as stubs (i.e. `class DebugProtocolHandler {};`.

File: vnext/JSI/Shared/ChakraRuntime.h (hunk 10 above)
Comment (JunielKatarn) [reply vnext/JSI/Shared/ChakraRuntime.h:329]: Yes!

**Review by NikoAri — APPROVED** [review]
:shipit:

**Review by NickGerleman — APPROVED** [review]
Signing off to backport.

**Pull request merged**